Allow parsing float-strings in bases 2 to 23

ChristianKurz opened the pull request:
This allows parsing Float-Strings in bases between 2 and 23. 
It doesn't allocate and is quite fast.

Fixes #22480.

Commit:
Fix #22480

Add feature to parse Floats in integer bases.
Detects index of '.' in s and uses Parse(Int,s).

Commit:
Update tests for issue #22480

Commit:
Update parse.jl

Commit:
Update parse.jl

Commit:
Update parse.jl

Commit:
Update parse.jl

Commit:
removed unnecessary 0

jw3126 commented on test/parse.jl:
```diff
@@ -901,3 +901,10 @@ let
+@test parse(Float64, "1.1", 2) == 1.5
+@test parse(Float64, "0.001", 2) == 0.125
+@test parse(Float64, "0.1", 3) == 0.3333333333333333
+@test_throws ArgumentError parse(Float64, "11", 2)
```
Why do you want that the number contains a point. Right now
```
julia> parse(Float64, "11")
11.0
```
works.

jw3126 commented on test/parse.jl:
```diff
@@ -901,3 +901,10 @@ let
+@test parse(Float64, "0.001", 2) == 0.125
+@test parse(Float64, "0.1", 3) == 0.3333333333333333
+@test_throws ArgumentError parse(Float64, "11", 2)
+@test_throws ArgumentError parse(Float64, "14.0", 3)
```
A lot more tests would be useful: 
* negative numbers, NaN, Inf..., 
* strings containing whitespace, scientific notation etc. 
* BigFloat, Float32 ...

jw3126 commented on base/parse.jl:
```diff
@@ -167,6 +167,29 @@ float(x::AbstractString) = parse(Float64,x)
 
 float{S<:AbstractString}(a::AbstractArray{S}) = map!(float, similar(a,typeof(float(0))), a)
 
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
```
A lot of edge cases are missing:
negative numbers, NaN, Inf, trailing whitespace etc.

jw3126 commented on base/parse.jl:
```diff
@@ -167,6 +167,29 @@ float(x::AbstractString) = parse(Float64,x)
+            continue
+        end
+        n += 1
+        d = parse(Int,s[i])
```
You can probably use `parse(Int, s[i], base)` here. This will work for even more bases and you don't need to check `d < base` yourself.

jw3126 commented on base/parse.jl:
```diff
@@ -167,6 +167,29 @@ float(x::AbstractString) = parse(Float64,x)
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
+    res = zero(T)
+    2 <= base <= 9 || throw(ArgumentError("invalid base: base must be 2 ≤ base ≤ 9, got $base"))
+    b = Float64(base)
```
`b = T(base)` ?

jw3126 commented on test/parse.jl:
```diff
@@ -901,3 +901,10 @@ let
+# issue #22480
+@test parse(Float64, "1.1", 2) == 1.5
+@test parse(Float64, "0.001", 2) == 0.125
+@test parse(Float64, "0.1", 3) == 0.3333333333333333
```
You should use `===` for such tests, to make sure that the types are correct.

shorty66 commented:
Thanks for those comments, i will see what i can do.

Commit:
Parse(Float,"",base): check edge cases, simplify

Function works with negative numbers, NaN and Inf

Used suggestions from comments

Commit:
updated tests for edge cases

Commit:
strip whitespace

As suggested strip leading and trailing whitespace. 
I'm not sure if it would be better to expect the user to take care of that:
This change now allocates memory to strip the string, even when no whitespaces are present.

Parsing the following array of 500.000 strings
```julia
julia> test = [string(rand([0,1],4)...,'.', rand([0,1],5)...) for i in 1:500000]
```
with whitespace stripping
```julia
julia> @benchmark parse.(Float64, test, 2)
BenchmarkTools.Trial:
  memory estimate:  34.33 MiB
  allocs estimate:  1000041
  --------------
  minimum time:     2.188 s (0.27% GC)
  median time:      2.215 s (0.16% GC)
  mean time:        2.282 s (0.19% GC)
  maximum time:     2.443 s (0.14% GC)
  --------------
  samples:          3
  evals/sample:     1
```

and without
```julia
julia> @benchmark parse.(Float64, test, 2)
BenchmarkTools.Trial:
  memory estimate:  3.82 MiB
  allocs estimate:  41
  --------------
  minimum time:     2.242 s (0.00% GC)
  median time:      2.389 s (0.00% GC)
  mean time:        2.369 s (0.00% GC)
  maximum time:     2.477 s (0.00% GC)
  --------------
  samples:          3
  evals/sample:     1
```

jw3126 commented on base/parse.jl:
```diff
@@ -167,6 +167,32 @@ float(x::AbstractString) = parse(Float64,x)
 float{S<:AbstractString}(a::AbstractArray{S}) = map!(float, similar(a,typeof(float(0))), a)
 
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
+    s = strip(s, ' ')
```
You should use `strip(s)` instead. Otherwise tabs/newlines etc. survive.

jw3126 replied:
Also this changes the type of `s` e.g. `String` -> `Substring` which is bad for performance.

jw3126 replied:
To resolve this you should probably reuse things from the `Integer` case like `parseint_preamble` for getting rid of whitespace.

shorty66 replied:
As i mentioned in the commit message:

I'm not sure if it would be better to expect the user to take care of strip():
This change now allocates memory to strip the string, even when no whitespaces are present.

Parsing the following array of 500.000 strings
```julia
julia> test = [string(rand([0,1],4)...,'.', rand([0,1],5)...) for i in 1:500000]
```
with whitespace stripping
```julia
julia> @benchmark parse.(Float64, test, 2)
BenchmarkTools.Trial:
  memory estimate:  34.33 MiB
  allocs estimate:  1000041
  --------------
  minimum time:     2.188 s (0.27% GC)
  median time:      2.215 s (0.16% GC)
  mean time:        2.282 s (0.19% GC)
  maximum time:     2.443 s (0.14% GC)
  --------------
  samples:          3
  evals/sample:     1
```

and without
```julia
julia> @benchmark parse.(Float64, test, 2)
BenchmarkTools.Trial:
  memory estimate:  3.82 MiB
  allocs estimate:  41
  --------------
  minimum time:     2.242 s (0.00% GC)
  median time:      2.389 s (0.00% GC)
  mean time:        2.369 s (0.00% GC)
  maximum time:     2.477 s (0.00% GC)
  --------------
  samples:          3
  evals/sample:     1
```

jw3126 replied:
I think the function should strip. And if you reuse the `Integer` case there will be no allocation.

jw3126 replied:
Interesting that the `strip` version is actually faster in your benchmark.

shorty66 replied:
These Benchmarks are not to be trusted - im running them on a CoreM CPU which throttles after a short boost period. I think, the second Benchmark had some throttled CPU-time.

jw3126 commented on base/parse.jl:
```diff
@@ -167,6 +167,32 @@ float(x::AbstractString) = parse(Float64,x)
+    sign = s[1] == '-' ? -1 : 1
+    sign == -1 && (s = s[2:end])
+    s == "NaN" && return T(NaN)
+    s == "Inf" && return T(Inf)
```
`-Inf`

Commit:
Allow "-Inf" to be parsed, remove tabs, newlines

Commit:
Check if -Inf is parsed correctly

Commit:
Merge branch 'master' into master

shorty66 commented:
I would like to go on with this PR. What is still needed? 
* @jw3126 mentioned tests for BigFloats, strings with whitespace, scientific notation.
* Stripping whitespace should be done similar to `parseint_preamble`.

Is that everything?

jw3126 commented:
Thanks for doing this. I think that plan sounds good!

stevengj commented on base/parse.jl:
```diff
@@ -240,6 +240,32 @@ float(x::AbstractString) = parse(Float64,x)
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
+    s = strip(s)
+    sign = s[1] == '-' ? -1 : 1
+    sign == -1 && (s = s[2:end])
```
need to handle a leading `+` too

stevengj replied:
Also, this is not type-stable, since slicing may not return the same type of string.  And this allocates a copy of the string — I think that `parse` should always be non-allocating.

stevengj commented on base/parse.jl:
```diff
@@ -240,6 +240,32 @@ float(x::AbstractString) = parse(Float64,x)
+    sign = s[1] == '-' ? -1 : 1
+    sign == -1 && (s = s[2:end])
+    s == "NaN" && return T(NaN)
+    s == "Inf" && return sign*T(Inf)
```
These comparisons should not be case-sensitive:
```jl
julia> parse(Float64, "nan")
NaN

julia> parse(Float64, "naN")
NaN
```

stevengj commented on base/parse.jl:
```diff
@@ -240,6 +240,32 @@ float(x::AbstractString) = parse(Float64,x)
+        n += 1
+        d = parse(Int,s[i], base)
+        res += d*b^(-n)
+    end
```
What about parsing numbers with explicit exponents, e.g. `"2.3e4"`?

ChristianKurz committed:
update julia

ChristianKurz committed:
add float parsing tests

includes tests for case sensitive NaNs and Infs

ChristianKurz committed:
added tests for explicit exponents in bases 4 & 5

ChristianKurz committed:
added tests for explicitly positive numbers and negative explicit exponents

ChristianKurz committed:
reimplemented float parsing in arbitrary bases

whitespace and explicit signs are taken care of with `parsefloat_preamble`

`inf` and `nan` are case insensitively checked and correctly handled

ChristianKurz committed:
added tests for trailing and leading whitespace

ChristianKurz committed:
made whitespace-tests  pass

shorty66 commented:
This now strips whitespace similar to parseint_preamble. Trailing whitespace is stripped in the main loops - this may be done nicer.
Tests are updated to reflect the new features.

ChristianKurz committed:
trim trailing whitespace, add test for BigFloats

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,69 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    sign*res*b^(expsign*exponent)
+end
+
+float(s::AbstractString, base::Integer) = parse(Float64,s,base)
```
Using `float` for parsing is deprecated. We should not add this method.

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,69 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    sign, c, i = parsefloat_preamble(s, base)
+
+    lowercase(c) == 'n' && return T(NaN)
+    lowercase(c) == 'i' && return sign*T(Inf)
```
With e.g. `base=50` these may be valid digits for finite numbers.

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,69 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    while !done(s, i)
+        c, i = next(s, i)
+        isspace(c) && break
+        c == 'e' && break
```
`e` may be a valid digit for e.g. a hex number. Maybe we should allow scientific notation only for bases smaller then `15`, where only digits `0,1,...a,b,c,d` can occur.

jw3126 reviewed (changes requested):
Other then the above, this looks very clean, nice!

ChristianKurz committed:
tests for explicit exponents without point, fixed bigfloat-test

I used `==` for the bigfloat test as using `===` wouldn't work. Is this expected?

ChristianKurz committed:
Only allow explicit exponents for base<16, fix nan, inf

This now detects nans and infs by all three letters. 
Explicit exponents are only allowed for bases under 16.
In addition, explicit exponents can now be used without a point.

Also removed `float()`-parsing, fixed type stability of `exponent` and reordered some operations to make the while-loops look alike.

ChristianKurz commented:
This now also uses the new keyword syntax for `parse(Int, "123", base=base)` and shouldn't fail the tests anymore.

ChristianKurz committed:
reduced allocation searching for inf and nan; trailing whitespace

ChristianKurz committed:
corrected scientific notation condition to <15 [ci skip]

ChristianKurz committed:
a second `parsefloat_preamble` slipped in; removed

ChristianKurz commented:
This is now even faster than `parse(Float64, A)` without the `base` variable. It does allocate quite a lot more memory, though - possibly due to the way `NaN` and `Inf` are handled.
My version allows for whitespace between `sign` and `figures`, but also silently overflows to `Inf` or `-Inf`.
The original version without base also does not parse strings ending in `e` and throws an error, while mine parses correctly.

```julia
base62 = vcat(string.(collect(0:9)),string.(collect('A':'Z')), string.(collect('a':'z')))
test(b,n=1) = [string(rand([" ", ""], rand(0:n))..., rand(["+", "-", ""]), rand(base62[1:b],rand(1:n))...,rand([".",""]), rand(base62[1:b], rand(0:n))...,rand([b<15?"e0":"",""]), rand(base62[1:b],rand(0:n))...) for i in 1:500000]
foo = test(10)
```
```julia
Julia> @benchmark parse.(Float64, foo)
BenchmarkTools.Trial:
  memory estimate:  3.82 MiB
  allocs estimate:  55
  --------------
  minimum time:     370.643 ms (0.00% GC)
  median time:      380.581 ms (0.00% GC)
  mean time:        380.223 ms (0.11% GC)
  maximum time:     392.700 ms (0.00% GC)
  --------------
  samples:          14
  evals/sample:     1

Julia> @benchmark parse.(Float64, foo, 10)
BenchmarkTools.Trial:
  memory estimate:  72.48 MiB
  allocs estimate:  749998
  --------------
  minimum time:     235.618 ms (3.03% GC)
  median time:      245.665 ms (3.01% GC)
  mean time:        265.533 ms (9.56% GC)
  maximum time:     422.304 ms (40.71% GC)
  --------------
  samples:          20
  evals/sample:     1
```

@stevengj, @jw3126 Would you mind merging this or are there any more things to correct?

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,82 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+        exponent = exponent * base + tmp
+    end
+
+    return sign*res*b^(expsign*exponent)
```
```julia
julia> parse(Float64, "1.2e2 some nonesense", 10)
120.0
```
Should throw an error instead, like here:
```
julia> parse(Float64, "1.2e2 some nonesense")
ERROR: ArgumentError: cannot parse "1.2e2 some nonesense" as Float64
```
Probably by checking that the rest of `s` is whitespace.

jw3126 replied:
Also 
```
julia> parse(Float64, "nanana")
ERROR: ArgumentError: cannot parse "nanana" as Float64
Stacktrace:
 [1] tryparse_internal(::Type{Float64}, ::String, ::Int64, ::Int64, ::Bool) at ./parse.jl:410
 [2] parse(::Type{Float64}, ::String) at ./parse.jl:417
 [3] top-level scope

julia> parse(Float64,  "nanana", 10)
NaN
```

jw3126 commented on test/parse.jl:
```diff
@@ -262,3 +262,48 @@ end
+
+@test parse(BigFloat, "2.3e-23",5) == big"2.129919999999999999999999999999999999999999999999999999999999999999999999999986e-09"
+
+@test_throws ArgumentError parse(Float64, "14.0", 3)
```
A test, where valid digits could be confused with scientific notation would be good.
```
@test parse(Float64, "1e3", 16) == 16^2 + 14*16 + 3
@test parse(Float64, "1e3", 10)  == 1000.
```

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,82 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+        end
+        all(nan) && return T(NaN)
+        all(inf) && return sign*T(Inf)
+    end
```
You could avoid these allocations by doing something like:
```julia
function _rest_is_whitespace(i,s)
    while !done(s,i)
        c,i = next(s,i)
        isspace(c) || return false
    end
    true
end
function _represents_nan_or_inf(i,s,xxx)
    length(s) < i+2 && return false
    for j in 1:3
        C,i = next(s,i)
        c = lowercase(C)
        c == xxx[j] || return false
    end
    _rest_is_whitespace(i,s)
end

function _represents_nan_or_inf(i,s)
    _represents_nan_or_inf(i,s,"nan") ||
    _represents_nan_or_inf(i,s,"inf")
end
```

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,82 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    return sign, startpos
+end
+
+function parse{T<:AbstractFloat}(::Type{T}, s::AbstractString, base::Integer)
```
`base` should be a keyword argument like in the integer case.

ChristianKurz committed:
restisspace & less memory intensive nan/inf parsing

ChristianKurz committed:
tests with keyword base, testing restisspace and nan/inf handling

ChristianKurz commented:
I have included your suggestions.
While memory usage went down a little, allocations have doubled. I couldn't find the reason for that - i'll try again tomorrow.

It came to my mind, that we should do something about parsing `NaN` and `Inf` if `base>23`. In that case `n` is allowed as digit and `NaN` could also mean a number.
Should we restrict parsing to `base<23` ?
Or ignore `Inf` and `NaN` if `base` is larger?

jw3126 commented:
I think it is reasonable to restrict to `base < 23`. I don't know if there is a usecase beyond `base=16`.

jw3126 commented on base/parse.jl:
```diff
@@ -217,6 +217,92 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+        c, i = next(s, i)
+        isspace(c) && break
+        c == '.' && break
+        if baseunder15 && c == 'e'
```
We should check for `lowercase(c) == 'e'`.
```julia
@test parse(Float64, "1E10", base=10) === 1E10
```
fails currently. Same below.

jw3126 commented:
On my machine this does not allocate anything:
```julia
julia> @benchmark parse(Float64, "1.123e34", base=10)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     299.047 ns (0.00% GC)
  median time:      304.661 ns (0.00% GC)
  mean time:        307.436 ns (0.00% GC)
  maximum time:     675.179 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     257
```

ChristianKurz committed:
lowercase e, base restricted to 2:23

reused `check_valid_base`

ChristianKurz committed:
test if explicit exponent is case-insensitive

jw3126 reviewed (approved):
LGTM

jw3126 commented:
I think actually instead of implementing `parse` directly, we should implement `tryparse`. Which is basically the same, but returns `nothing` instead of an error for invalid input. This would allow to delete the existing `tryparse` [code](https://github.com/ChristianKurz/julia/blob/aa7da27a3ebd705fc374d86d5bd435ade685f899/base/parse.jl#L308) for various float types.

ChristianKurz commented:
I also tried to remove the other float-parsing functions but came to the same conclusion.
We could do something about that in a seperate PR.

Regarding this PR: 
I think this could be merged. I tried seperating the logic in several functions, but that doesn't increase readability that much and seemed like code-churn.
Should i add a NEWS entry?

jw3126 commented:
I think a NEWS entry does not hurt.

jw3126 commented:
I also think this is ready to be merged.

ChristianKurz committed:
Merge branch 'master' into master

ChristianKurz committed:
added entry about parsing floats in bases 2 to 23

ChristianKurz committed:
new entry should rather be under "library improvements"

stevengj commented on base/parse.jl:
```diff
@@ -218,6 +218,93 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    c = ' '
+    sign = 1
+    while isspace(c)
+        c, i = next(s,i)
```
This is abusing the iteration protocol.  It is assuming that `start(s)` is the same as the starting index `i=1`.

Better to use `i = nextind(s,i); c = @inbounds s[i]`

stevengj replied:
See also #25833.

You should also initialize `i` to `firstindex(s)`.

ChristianKurz replied:
This is also true for `parseint_preamble` where i drew inspiration from.

Should i update only the code from this PR to not abuse the iteration protocol?
I could also just leave it like this and make a seperate PR to update the whole `parse.jl` file.

stevengj replied:
@Keno, do you have a preference on how to proceed here?

ChristianKurz replied:
I replaced `next()` and `done()` calls in the whole file - this shouldn't 'abuse' the iteration protocol any longer.

stevengj commented on base/parse.jl:
```diff
@@ -218,6 +218,93 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
+    isempty(s) && throw(ArgumentError("Empty string!"))
+    c = ' '
+    sign = 1
+    while isspace(c)
```
Not that this loop will be a problem if `s` consists entirely of spaces.  (You'll get a `BoundsError`.)

ChristianKurz replied:
got it, will add a `!done(i,s)` condition.

ChristianKurz committed:
initialize i as firstindex(s), check

stevengj commented on base/parse.jl:
```diff
@@ -218,20 +218,22 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
 end
 
 ## string to float functions ##
-function parsefloat_preamble(s::AbstractString, base::Int, i::Int=1)
```
You can just use `i::Int=firstindex(s)` if you want to keep this as an optional argument.

ChristianKurz replied:
i actually didn´t see a need for the optional argument to an internal function.

Using `startpos` and `endpos` may be useful to eg. strip currency symbols. But i think this is orthogonal to this PR.

ChristianKurz committed:
Merge branch 'master' of https://github.com/JuliaLang/julia

ChristianKurz committed:
do not abuse iteration protocol

replaced
`c, i = next(s,i)` with `c = @inbounds s[i]; i = nextind(s,i)` and
`done(s, i)` with i > lastindex(s)`

ChristianKurz committed:
added testset

ChristianKurz commented:
Rebased, removed abusive use of iteration protocol.
Tests pass locally.

ChristianKurz committed:
fixed trailing whitespace

ChristianKurz commented:
CI-Failures are unrelated - `parse.jl` tests do pass.

StefanKarpinski commented on base/parse.jl:
```diff
@@ -46,8 +46,9 @@ end
-    c, startpos = next(s,startpos)
-    c, startpos, j
+    c = @inbounds s[startpos]
+    startpos = nextind(s, startpos)
```
I know this was in the old code too, but just to check: we know that this is inbounds because `startpos <= endpos`, right?

fredrikekre commented:
> CI-Failures are unrelated - `parse.jl` tests do pass.

Looking at the logs I would say they are definitely related. For example:
```
parse(Float64, SubString("10", 1, 1)) === 1.0
  ArgumentError: input string is empty or only contains whitespace
```

ChristianKurz committed:
Rebased parse.jl

ChristianKurz committed:
parse only bases other than 10 with new function

ChristianKurz committed:
resolve conflicts

vtjnash commented:
I believe it will not be this simple. x-ref the [Parsers.jl package](https://github.com/JuliaData/Parsers.jl/blob/main/src/floats.jl) @quinnj
